---
title: "Visualizing Epigenomic Big Data Using genomation and Gviz R/Bioconductor packages"
author: "David Chen (github.com/ydavidchen)"
date: "12/14/2016"
output: html_document
---

## Overview

In this tutorial, we will practice visualization of two types of high-throughput epigenomics data: DNA methylation and chromatin modification. The **Roadmap Epigenomics** project provides high-quality data sets for our data visualization exercise. Several such data sets have been collected on the **CD3 primary cells**:

* DNA methylation by Reduced Representation Bisulfite Sequencing (RRBS)
* Chromatin modification (specifically, H3K27ac, collected by chromatin immuno precipitation followed by deep sequencing, ChIP-seq)

To learn more about these biotechnologies, please refer to **Park PJ. Nat. Rev.Genet. 10 (2009)** and **Bock C Nat. Rev. Genet. 13 (2012)**. 

To be able to follow along this tutorial, you should have a strong grasp of the R programming language and experience with Bioconductor tools. You should also be familiar with `GenomicRanges` operations. Linux/Unix is not required for this tutorial.

## Data Source

You may access the Roadmap Epigenomics metadata (a Google spreadsheet) to identify samples. Here, we see that files labeled $E33$ or $E34$ are from the CD3 cell line:

```{r metadata, eval=FALSE}
# install.packages("gsheet")
library(gsheet)
metaData <- gsheet2tbl("https://docs.google.com/spreadsheet/ccc?key=0Am6FxqAtrFDwdHU1UC13ZUxKYy1XVEJPUzV6MEtQOXc&usp=sharing#gid=15")
```

Download the CD3 data sets directly from **Roadmap Epigenomics** web portal. Alternatively, you can manually fetch those files from [Roadmap Epigenomics](http://egg2.wustl.edu/roadmap/web_portal/) or [Gene Expression Omnibus](https://www.ncbi.nlm.nih.gov/geo/roadmap/epigenomics/) and then load into R using `rtracklayer::import`.

```{r WebDataSets, message=FALSE}
# source("https://bioconductor.org/biocLite.R")
# biocLite(c("rtracklayer", "GenomicRanges"))
library(rtracklayer)
library(GenomicRanges)

## Speed up computation by using multiple cores:
# install.packages("doParallel")
library(doParallel)
registerDoParallel(detectCores() - 1)

## Download & import RRBS methylation data set as GRanges object:
CD3_rrbs_path <- "http://egg2.wustl.edu/roadmap/data/byDataType/dnamethylation/RRBS/FractionalMethylation_bigwig/E033_RRBS_FractionalMethylation.bigwig"
CD3_rrbs <- import.bw(CD3_rrbs_path)
CD3_rrbs

## Download & import histone .broadPeak data as GRanges object (standard approach):
CD3_H3K27ac_path <- "http://egg2.wustl.edu/roadmap/data/byFileType/peaks/consolidated/broadPeak/E034-H3K27ac.broadPeak.gz"

## Select columns (if your data file to read is narrowPeak, add `peak = "integer"` to the vector):
extraCols <- c(singnalValue="numeric", pValue="numeric", qValue="numeric")
CD3_broadPeak <- import(CD3_H3K27ac_path, format="BED", extraCols=extraCols)
CD3_broadPeak
```

As an exercise, we will be examining the chromatin and DNA-methylation landscape on chromosome X (chrX). 

```{r CommonColumns, message=FALSE}
CD3_rrbs <- CD3_rrbs[seqnames(CD3_rrbs) == "chrX"]
CD3_broadPeak <- CD3_broadPeak[seqnames(CD3_broadPeak) == "chrX"]

CD3_rrbs@elementMetadata$CommonCol <- CD3_rrbs@elementMetadata$score #common column
CD3_rrbs@elementMetadata$score <- NULL
CD3_rrbs #preview

CD3_broadPeak@elementMetadata$CommonCol <- CD3_broadPeak@elementMetadata$singnalValue #common column
CD3_broadPeak@elementMetadata$singnalValue <- NULL
CD3_broadPeak@elementMetadata$name <- NULL
CD3_broadPeak@elementMetadata$score <- NULL
CD3_broadPeak@elementMetadata$pValue <- NULL
CD3_broadPeak@elementMetadata$qValue <-NULL
CD3_broadPeak #preview
```

Define transcription start using Bioconductor's annotation package, which is based on the UCSC hg19 database. Note that if you prefer to fetch the annotation using a different method (e.g. `AnnotationHub` in R) or manually from the website, you are welcome to do so.

```{r Promoters, message=FALSE}
# source("https://bioconductor.org/biocLite.R")
# biocLite("TxDb.Hsapiens.UCSC.hg19.knownGene")
library(TxDb.Hsapiens.UCSC.hg19.knownGene) #annotation package
myPromoters <- promoters(TxDb.Hsapiens.UCSC.hg19.knownGene, upstream=500, downstream=500) #adjust as necessary
```

**In case you are unable to download the data sets above, I've provided you with an R object with everything up to this point**

```{r eval=FALSE}
load("my-R-objects-for-genomation.RData") #objects so far
```

## Genome-scale data visualization by genomation

Use the genomation package to plot a heat matrix for the RRBS 

```{r ScoreMatrices, message=FALSE}
# source("https://bioconductor.org/biocLite.R")
# biocLite("genomation")
library(genomation)
library(GenomicRanges)

## Define colors:
# install.packages("RColorBrewer")
library(RColorBrewer)
myCols <- brewer.pal(9, "Set1")

mySMlist <- ScoreMatrixList(
  targets = list(RRBS=CD3_rrbs, H3K27ac=CD3_broadPeak), 
  windows = myPromoters,
  strand.aware = TRUE, 
  weight.col = "CommonCol",
  is.noCovNA = TRUE #REQUIRED for data without complete coverage, like methylation
)

## Heat matrix for RRBS data:
heatMatrix(mySMlist[[1]], xcoords=c(-500, 500), xlab="Distance to transcription start (bp)", main = "chrX 5mC methylation by RRBS in CD3")

## Genomic Curve for H3K27ac data:
plotMeta(mySMlist[[2]], line.col=myCols[2], xcoords=c(-500, 500), xlab="Distance to transcription start (bp)", main="chrX H3K27ac by ChIP-seq in CD3")
```

## Single-gene-level data visualization using `Gviz`

Gviz is a `grid`-based package that generates publication-quality genomic plots. In this tutorial, we will examine the use of Gviz for visualizing H3K27ac and DNA methylation at two adjacent loci on the X chromosome

```{r Gviz}
library(Gviz)
myFavoriteGene <- c("EIF2S3", "KLHL15") #c("GENE1", "GENE2", ...)
gen <- "hg19" #genome build
chr <- "chrX"
```

Download gene and chromosome annotations from the UCSC hg19 database:

```{r UCSC, message=FALSE}
mySession <- browserSession('UCSC')
genome(mySession) <- gen
query <- ucscTableQuery(mySession,track='refGene')
tableName(query) <- "refGene"
refseq <- getTable(query)

## Subset your tracks and tables as much as you can to save time and minimize RStudio crashing:
refseq <- refseq[refseq$name2 == myFavoriteGene, ] #further subset
TSS <- min(refseq$txStart)
from <- TSS - 5000 #add 5kb extra 
to <- max(refseq$cdsEnd) + 5000 #add 5kb extra 
paste0(chr,":",from,"-",to) #can be searched on UCSC Genome Broswer
```

Make track objects for chromosome band and genomic axis:

```{r chromTracks}
ideoTrack <- IdeogramTrack(genome=gen, chromosome=chr, from=from, to=to)
axTrack <- GenomeAxisTrack(genome=gen, chromosome=chr, from=from, to=to)
```

Create a gene-region track using the RefSeq annotation:

```{r GeneModel, message=FALSE}
GeneTrack <- UcscTrack(
  genome=gen, chromosome=chr, 
  track="refGene", trackType = "GeneRegionTrack",
  from=from, to=to,
  rstarts="exonStarts", rends="exonEnds", strand="strand",
  gene="name", symbol="name2", transcript="name", 
  transcriptAnnotation="symbol", #labels gene names
  stacking="dense", #options: "full", "pack", & "dense" as in UCSC Genome Browser
  fill="salmon", col="salmon", name="RefSeq Genes" #update based on your choice
)
```

Use your data sets to make `DataTrack` objects:

```{r DataTracks}
dTrack_RRBS <- DataTrack(
  CD3_rrbs,
  type="histogram", 
  genome=gen, chromosome=chr,
  from=from, to=to,
  fill="blue", col="blue",
  name="RRBS"
)
dTrack_RRBS

dTrack_H3K27ac <- DataTrack(
  CD3_broadPeak,
  type="histogram", 
  genome=gen, chromosome=chr,
  from=from, to=to,
  fill="orange", col="orange",
  name="H3K27ac"
)
dTrack_H3K27ac
```

You may include additional tracks that represent (epi)genomic information. The syntax is identical to the gene-region track:

```{r Additional, message=FALSE}
cpgIslands <- UcscTrack(
  genome=gen, chromosome=chr, track="cpgIslandExt", 
  from=from, to=to,
  trackType="AnnotationTrack", 
  start="chromStart",
  end="chromEnd", 
  id="name", shape="box",
  stacking="dense",
  fill="#006400", col="#006400", name="CpG islands"
)
```

You are now ready to plot the track objects:

```{r GvizPlot}
plotTracks(
  c(ideoTrack, axTrack, GeneTrack, cpgIslands, dTrack_RRBS, dTrack_H3K27ac),
  genome=gen, chromosome=chr, 
  reverseStrand=FALSE,
  from=from, to=to,
  background.title = "black"
)
```

A separate tutorial is written on how to visualize methylation array data.

## Summary

If you plan to revisit your plots, it might be a good idea to save your workspace. As you may have seen, generating these plots is computationally intensive and therefore time consuming. 

```{r Save, eval=FALSE}
storagePath <- "" #Type your path here
save(list=ls(all.names=TRUE), file=paste(storagePath, "my-genomic-data-vis-workspace.RData", sep="/"))
```

## References

**Akalin, A. et al.** genomation: a toolkit to summarize, annotate, and visualize genomic intervals. Bioinfo 31(7): 1127-1129 (2015).

**Bock, C.** Analyzing and interpreting DNA methylation Data. Nature Rev. Genet. 13: 705-719 (2012).

**Hahne F, Ivanek R.** Visualizing genomic data using Gviz and Bioconductor. Methods Mol. Biol. 1418: 335-351 (2016)

**Park, PJ.** ChIP-seq: advantages and challenges of a maturing technology. Nature Rev. Genet. 10: 669-680 (2009).

**Roadmap Epigenomic Consortium, et al.** Integrative analysis of 111 reference human epigenomes. Nature 518: 317-330 (2015). 