---
title: "Visualization of Illumina HumanMethylationEPIC (850K) data set using the Gviz R/Bioconductor package"
author: "David Chen (github.com/ydavidchen)"
date: "12/14/2016"
output: html_document
---

## Introduction

Gviz is an intricate tool for visualizing (epi)genomic annotation and data. In this tutorial, a DNA 5-methylcytosine (5mC) methylation levels from a simulated data set from the ChAMP package. 

## Workflow for visualizing DNA methylation datasets

For first time users, be sure to download the appropriate annotation package, depending on your methylation data:

* (HumanMethylation450k annotation)[http://bioconductor.org/packages/release/data/annotation/html/IlluminaHumanMethylation450kanno.ilmn12.hg19.html]
* (HumanMethylation850k (EPIC) annotation)[http://bioconductor.org/packages/release/data/annotation/html/IlluminaHumanMethylationEPICanno.ilm10b2.hg19.html]

You also have the option of manually loading the Illumina 850K annotation package into R. Visit Illumina's website on (DNA methylation)[http://www.illumina.com/techniques/microarrays/methylation-arrays.html] for more information and downloads.

```{r Annotation, message=FALSE}
## Load packages:
library(Gviz)
library(rtracklayer)

## Speed up computation by using all minus 1 cores:
library(doParallel); registerDoParallel(detectCores() - 1)

## Load 850K annotation package
library(IlluminaHumanMethylationEPICanno.ilm10b2.hg19) 
data(IlluminaHumanMethylationEPICanno.ilm10b2.hg19)
annot.850k <- getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b2.hg19)
annot.850k <- as.data.frame(annot.850k@listData)
```

In the following example, we will focus on the $HOXD1$ gene. 

```{r GeneSubset}
## Subset CpGs located in your favorite gene(s):
CpGinfo <- data.frame(
  cgID = annot.850k$Name,
  GeneName = annot.850k$UCSC_RefGene_Name,
  GenomicContext = annot.850k$UCSC_RefGene_Group,
  TranscriptIsoforms = annot.850k$UCSC_RefGene_Accession
)

myFavoriteGene <- "HOXD1" #c("GENE1", "GENE2", ...) if multiple
CpGinfo <- CpGinfo[grep(myFavoriteGene, CpGinfo$GeneName, ignore.case=TRUE), ]

## Select CpG IDs within your favorite gene(s) based on genomic context:
myCpGs <- droplevels(CpGinfo$cgID)
myCpGs

## Define genome build:
gen <- "hg19"

## Select chromosome where your favorite gene resides:
chr <- "chr2" #here: HOX2D
```

Fetch genomic annotations from the UCSC hg19 database:

```{r UCSC, message=FALSE}
mySession <- browserSession('UCSC')
genome(mySession) <- gen
query <- ucscTableQuery(mySession,track='refGene')
tableName(query) <- "refGene"
refseq <- getTable(query)

## RECOMMENDED: Subset your tracks and tables as much as you can to save time and minimize RStudio crashing:
refseq <- refseq[refseq$name2 == myFavoriteGene, ] #further subset; use %in% if multiple
TSS <- min(refseq$txStart)
from <- TSS - 5000 #define start of plotting
to <- max(refseq$cdsEnd) + 5000 #define end of plotting
paste0(chr,":",from,"-",to) #searchable on UCSC Genome Broswer
```

Create chromosome ideograms and genome axis as separate tracks:

```{r chromTracks}
ideoTrack <- IdeogramTrack(genome=gen, chromosome=chr, from=from, to=to)
axTrack <- GenomeAxisTrack(genome=gen, chromosome=chr, from=from, to=to)
```

Create a gene-region track that represents the UCSC gene model:

```{r GeneModel, message=FALSE}
GeneTrack <- UcscTrack(
  genome=gen, chromosome=chr, track="refGene",
  trackType = "GeneRegionTrack",
  from=from, to=to,
  rstarts="exonStarts",rends="exonEnds", strand="strand",
  gene="name", symbol="name2", transcript="name", transcriptAnnotation="symbol",
  stacking = "dense", #options: "full", "pack", & "dense" as in UCSC Genome Browser
  fill = "salmon", col="salmon", name = "RefSeq Genes"
)
GeneTrack
```

Load Infinium 850k example dataset that comes with the `ChAMP` R/Bioconductor package:

```{r SimData, message=FALSE}
library(ChAMP); data(EPICSimData)
myLoad <- myLoad$beta
head(myLoad, 3)
```

When performing this procedure with your own data set, be sure to save a matrix of methylation beta values. This matrix can be exported from Illumina's GenomeStudio software. If you are working with Intensity Data Files (.idat) in R, you may use the function `getBeta` from the `minfi` package and then save your file using `write.table` or equivalent. Inclusion of multiple samples is supported by `Gviz`.

First, make an Illumina 850K `AnnotationTrack`:

```{r Setup, message=FALSE}
## Create (subset) data.frame for making GRanges object for methylation array annotation track AND data tracks:
df.850k <- data.frame(
  chr = annot.850k$chr,
  start = as.numeric(annot.850k$pos),
  end = as.numeric(annot.850k$pos),
  cgID = annot.850k$Name,
  UCSCgene = annot.850k$UCSC_RefGene_Name
)
rownames(df.850k) <- df.850k$cgID #for merging

## Make GRanges object:
gr.850k <- makeGRangesFromDataFrame(df.850k, keep.extra.columns=FALSE)
gr.850k <- gr.850k[seqnames(gr.850k) == chr, ] #subset to prevent RStudio crashing

## Make an Illumina 850k AnnotationTrack:
Annotation850K <- AnnotationTrack(
  gr.850k,
  chromosome=chr, genome=gen,
  start=from, end=to,
  stacking="dense",
  name="850K", fill="hotpink", col="hotpink")
Annotation850K 
```

We can then create a `DataTrack` object using the function `Gviz::DataTrack`. We will use it for plotting at the end:

```{r DataTrack}
## Create GRanges object using the subsetted 850k annotation:
gr.myLoad <- merge(df.850k, myLoad, by="row.names")
gr.myLoad <- makeGRangesFromDataFrame(gr.myLoad, keep.extra.columns=TRUE)

## DataTrack object is created using the GRanges object:
dTrack <- DataTrack(gr.myLoad,
                    type = "histogram", 
                    genome=gen, chromosome=chr, 
                    from=from, to=to, name="fraction of methyl.")

dTrack #preview track info, esp. number of samples
```

You may include additional tracks that represent (epi)genomic information. `Gviz::AnnotationTrack` and `Gviz::UcscTrack` are useful functions for this purpose. Here, we will visualize CpG Islands and common SNPs:

```{r Additional, message=FALSE}
## Fetch CpG island annotation from UCSC and turn it into a track object:
CpGislands <- UcscTrack(
  genome=gen, chromosome=chr, track="cpgIslandExt", 
  from=from, to=to,
  trackType = "AnnotationTrack", 
  start = "chromStart", end = "chromEnd",
  id="name", shape="box",
  stacking="dense", 
  fill="#006400", col="#006400", name="CGI"
)
CpGislands

## Fetch SNP annotation from UCSC and turn it into a track object:
CommonSNPs <- UcscTrack(
  genome=gen, chromosome=chr, track="Common SNPs(138)",
  from=from, to=to,
  trackType = "AnnotationTrack",
  start="chromStart", end="chromEnd", id="name",
  feature="func", strand="strand", shape="box",
  stacking="dense", 
  fill="purple", col="purple", name="SNP"
)
CommonSNPs 
```

You are now ready to visulaize your tracks. The plan is to highlight just the promoter region of the gene, and then plot a window with features defined above with a highlighting box:

```{r Plotting, message=FALSE}
plotTracks(
  c(ideoTrack, axTrack, GeneTrack, CpGislands, CommonSNPs, Annotation850K, dTrack),
  genome=gen, chromosome=chr, 
  reverseStrand = FALSE,
  from=from, to=to,
  background.panel="#FFFEDB", background.title="darkblue"
)
```

You can take a look at a close-up view:

```{r}
plotTracks(
  c(ideoTrack, axTrack, GeneTrack, CpGislands, CommonSNPs, Annotation850K, dTrack),
  genome=gen, chromosome=chr, 
  reverseStrand = FALSE,
  from=from-6000, to=from-3000, #promoter region
  background.panel="#FFFEDB", background.title="darkblue"
)
```

If you have generated a `Gviz` plots using the code above, congratulations! This is not an easy task. As you may have noticed, delicate graphics in `Gviz` require intense computation. You may save your work space as an `.RData` file. This will save you time if you want to revisit your `Gviz` objects at a later time:

```{r, eval=FALSE}
## RECOMMENDED: Save workspace for easy loading later:
storagePath <- "" #path, such as the Downloads folder
save(list=ls(), file=storagePath, envir=.GlobalEnv)
```

## References

Hahne F and Ivanek R. Visualizing genomic data using Gviz and Bioconductor. Methods Mol. Biol. 1418: 335-351 (2016)

Morris TJ, et al. ChAMP: 450k Chip Analysis Methylation Pipeline. Bioinfo. 30(3): 428-430 (2014).
